core/src/main/scala/org/apache/spark/Partition.scala

```diff
@@ -26,6 +26,11 @@ trait Partition extends Serializable {
    */
   def index: Int
 
+  override def equals(other: Any): Boolean = other match {
+    case that: Partition => super.equals(that) && that.index == index
```

Comment:
Hm, `super.equals()` delegates to the default in `Object` which requires reference equality. I don't think we can have that. Although defining these in an abstract class is dicey, I agree it should go hand in hand with `hashCode` at least and should just define equality based on index.